[typescript] Generic date type (#1966)
mui-org/material-ui-pickers: docs/pages/guides/typescript.mdx

import PageMeta from '_shared/PageMeta';

<PageMeta
  title="Typescript – @material-ui/pickers"
  description="A guide of how to use @material-ui/pickers with typescript"
/>

> ## Typescript guide

Reviewer: As far as I know, these capitalizations are not correct: "Typescript", "typescript", the only valid one is "TypeScript": http://typescriptlang.org/.

```suggestion
## TypeScript guide
```

Reviewer: cc @dtassone I have seen the same in the data grid.

> ## TypeScript guide

Reviewer: ```suggestion
# TypeScript guide
```

Author: Ohhh no no no, our styling is really weird if using right h1, h2 :) 

Needs to be fixed or wait for migration

Reviewer: Ok, let's wait migration :)

> Pickers are originally written in typescript, so type definetions are provided with the code without any additional requirements.

Reviewer: It don't see the link between written in TypeScript and not required additional requirement. What about:

```suggestion
The date picker components come with built-in TypeScript definitions.
```

Here is some information about how to achive better experience with typescript.

<!-- https://stackoverflow.com/questions/62452160/material-datetimepicker-onchange-typescript-mismatch-error/62471345 -->

> #### Nullable value

Reviewer: The heading structure needs to be linear h1 > h2 > h3, jumps are not allowed

```suggestion
### Nullable value
```

Any picker can return null in the `onChange` argument so if you are using `React.useState`, you **must** manually cast your default value as nullable. (e.g. `Date | null`)

```tsx

> const MyComponent = () => {

Reviewer: ```suggestion
function MyComponent() {
```

>   const [date, setDate] = React.useState<Date | null>(new Date());

Reviewer: Assuming we don't know what the value will be used for, might be best to keep it generic?

```suggestion
  const [value, setValue] = React.useState<Date | null>(new Date());
```

  return (
    <DatePicker
      value={date}

>       onChange={date => setDate(date)}

Reviewer: It might be best to avoid the implicit returned value, I have already been leaning toward it to ease refactorization, but it doesn't make a huge different

```suggestion
      onChange={date => { setDate(date) }}
```

      renderInput={props => <TextField {...props} />}
    />
  );
};
```

#### Generics

> We don't really know which adapter you will pass to the `<LocalizationProvider />`. So any picker component is generic by default.

Reviewer: Please remove all usage of "we" for "it" when mentioning how the code behaves.
```suggestion
The components don't know which adapter you will pass to the `<LocalizationProvider />`. So any picker component is generic by default.
```

Reviewer: cc @mbrookes has this guideline come from him
